kv: allow cancellation of commands in Replica.Send

spencerkimball opened the pull request:
In order to return control to the caller on context cancellation,
this change moves the processing done by `Replica.EndCmds` to post-Raft
execution. In the event of cancellation, the caller receives an
`AmbiguousResultError` and can continue, while the command queue
remains blocked until Raft finishes processing the command, after
which the timestamp cache can also be updated.

Fixes #10488
Fixes #10733

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10772)
<!-- Reviewable:end -->

spencerkimball committed:
kv: allow cancellation of commands in Replica.Send

In order to return control to the caller on context cancellation,
this change moves the processing done by `Replica.EndCmds` to post-Raft
execution. In the event of cancellation, the caller receives an
`AmbiguousResultError` and can continue, while the command queue
remains blocked until Raft finishes processing the command, after
which the timestamp cache can also be updated.

Fixes #10488
Fixes #10733

tamird commented:
Reviewed 2 of 5 files at r1.
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1236 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnYnKQ4btCI_-5pz6w:-KWnYnKQ4btCI_-5pz6x:b-eaj5wc) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L1236)):*
> ```Go
> // returned error replaces the supplied error.
> func (ec *endCmds) done(br *roachpb.BatchResponse, pErr *roachpb.Error, shouldRetry bool) {
> 	if ec == nil {
> ```

not a fan of this pattern in general. can it be pushed up to the caller(s)?

---

*[pkg/storage/replica.go, line 3075 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnYbpoTZTF06g-AJ5v:-KWnYbpoTZTF06g-AJ5w:brz4hwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L3075)):*
> ```Go
> 				ctx = nil
> 				cmd.sendDone(proposalResult{ShouldRetry: true})
> 			}(*cmd)
> ```

this dereference, coupled with the overwriting of `endCmds` below is subtle.

---

*[pkg/storage/replica.go, line 3078 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnYFIBhXiM6LbcYFG5:-KWnYFIBhXiM6LbcYFG6:bomasfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L3078)):*
> ```Go
> 			// Clear the endCmds and doneCh so that when
> 			// ProposalData.sendDone() is a noop when invoked below.
> 			cmd.endCmds = nil
> ```

what is the difference between `cmd.endCmds` and `cmd.LocalProposalData.endCmds`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1224 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnXaDKLT5R4PC0OnwC:-KWnXaDKLT5R4PC0OnwD:b4mezmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L1224)):*
> ```Go
> 
> // endCmds holds necessary information to end a batch after Raft command
> // processing: global and local command queue
> ```

`s/queue/queue./g` Feels like the sentence ends abruptly too.

---

*[pkg/storage/replica.go, line 1278 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnXs9VOA5LBSpMjwwV:-KWnXs9VOA5LBSpMjwwW:buri5au) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L1278)):*
> ```Go
> 	ec.repl.cmdQMu.local.remove(ec.cmdLocal)
> 	ec.repl.cmdQMu.Unlock()
> }
> ```

I assume this is code movement. If there was anything substantial that changed here, please point it out.

---

*[pkg/storage/replica.go, line 3076 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnZLlyz7ZsMlIGxq4z:-KWnZLlyz7ZsMlIGxq5-:bv20qfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L3076)):*
> ```Go
> 				cmd.sendDone(proposalResult{ShouldRetry: true})
> 			}(*cmd)
> 			// Clear the endCmds and doneCh so that when
> ```

`s/that when/that/g`

---

*[pkg/storage/replica_proposal.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnYkJQiLniAHdwWa0X:-KWnYkJQiLniAHdwWa0Y:b24rjy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica_proposal.go#L104)):*
> ```Go
> // and possibly no longer listening to this done channel, and so can't
> // be counted on to invoke endCmds itself.
> func (lpd *LocalProposalData) sendDone(pr proposalResult) {
> ```

I would have named this method `finish`, but I don't feel strongly about this.

---

*[pkg/storage/replica_proposal.go, line 105 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnYui84j9l3e_EiZ_S:-KWnYui84j9l3e_EiZ_T:b8rtlro) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica_proposal.go#L105)):*
> ```Go
> // be counted on to invoke endCmds itself.
> func (lpd *LocalProposalData) sendDone(pr proposalResult) {
> 	lpd.endCmds.done(pr.Reply, pr.Err, pr.ShouldRetry)
> ```

Should you set `lpd.endCmds = nil` here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-:-KWnZqAcHaudQs-4Cev9:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 2 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1224 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnXaDKLT5R4PC0OnwC:-KWn_2yeVPLBsQzh-w7_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L1224)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/queue/queue./g` Feels like the sentence ends abruptly too.
</details>
Done.

---

*[pkg/storage/replica.go, line 1236 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnYnKQ4btCI_-5pz6w:-KWn_RNE2lcH_Nq1_jZT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L1236)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not a fan of this pattern in general. can it be pushed up to the caller(s)?
</details>
Done.

---

*[pkg/storage/replica.go, line 1278 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnXs9VOA5LBSpMjwwV:-KWn_TZNB1xZ5az5rr_A:b-7ka5ei) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L1278)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I assume this is code movement. If there was anything substantial that changed here, please point it out.
</details>
Nothing.

---

*[pkg/storage/replica.go, line 3075 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnYbpoTZTF06g-AJ5v:-KWnaNeWddj74KgDQCyW:bxp6gum) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L3075)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this dereference, coupled with the overwriting of `endCmds` below is subtle.
</details>
I made it less subtle

---

*[pkg/storage/replica.go, line 3076 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnZLlyz7ZsMlIGxq4z:-KWn_h0A59RZHd97zLr_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L3076)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/that when/that/g`
</details>
Done.

---

*[pkg/storage/replica.go, line 3078 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnYFIBhXiM6LbcYFG5:-KWn_Xs04fIy0Tee2MOg:bg4gxj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica.go#L3078)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what is the difference between `cmd.endCmds` and `cmd.LocalProposalData.endCmds`?
</details>
Nothing. `LocalProposalData` is an embedded struct. I guess Tobi was just being explicit here; I followed the existing patterns.

---

*[pkg/storage/replica_proposal.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnYkJQiLniAHdwWa0X:-KWnaaZiiAhmeax2fasG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica_proposal.go#L104)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I would have named this method `finish`, but I don't feel strongly about this.
</details>
Done.

---

*[pkg/storage/replica_proposal.go, line 105 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWnYui84j9l3e_EiZ_S:-KWnaavHIJn3qhPyIoEr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b405e726dbd2d737e0ab0d617b8e39ec568334/pkg/storage/replica_proposal.go#L105)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Should you set `lpd.endCmds = nil` here?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1235 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoMfmnHNtnphKvV4Q6:-KWoMfmoW4bxAkvWnCWO:by6hj8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L1235)):*
> ```Go
> // the timestamp cache using the final timestamp of each command.  The
> // returned error replaces the supplied error.
> func (ec *endCmds) done(br *roachpb.BatchResponse, pErr *roachpb.Error, shouldRetry bool) {
> ```

The amount of code movement here makes it very hard to review, and also needless acquires the `git blame` on a much larger-than-necessary swath of code. In the future, please consider leaving this code where it is; in this case, putting all of `endCmds` after the definition of `beginCmds` would have made things much more palatable and clear.

---

*[pkg/storage/replica.go, line 1800 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoNCmjK6qOiZPOVOal:-KWoNCmjK6qOiZPOVOam:bfu0jkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L1800)):*
> ```Go
> 		select {
> 		case propResult := <-ch:
> 			endCmds = nil // these will have been invoked post-Raft.
> ```

this comment is anemic, can you flesh it out? In particular, isn't it always the intention that `endCmds` is processed post-Raft?

---

*[pkg/storage/replica.go, line 1805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoNmzEruMqPl8UMkmY:-KWoNmzEruMqPl8UMkmZ:b-7xtfeb) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L1805)):*
> ```Go
> 			// If our context was cancelled, return an AmbiguousResultError
> 			// to indicate to caller that the command may have executed.
> 			// However, we proceed only if the command isn't already being
> ```

this comment is confusing to put it mildly.

The ambiguous result error is only returned if the context was cancelled **and** the command might already be executing, right? The first sentence states that the ambiguous result is returned unconditionally, and then the second sentence adds a condition.

Also, if the context is cancelled and then we fail to abandon, this loops back around and waits for the result of the raft command, even though the client no longer cares - is that the intention?

It seems that `endCmds` is being set to nil below to avoid the defer, but again, it deserves explicit commentary. In fact, I would rather we removed that defer completely; the few lines saved are not worth the subtlety and confusion that result from its presence.

---

*[pkg/storage/replica.go, line 1817 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoPEWJnbBVgNDk5FAO:-KWoPEWJnbBVgNDk5FAP:bt5aezq) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L1817)):*
> ```Go
> 			// If we're shutting down, return an AmbiguousResultError to
> 			// indicate to caller that the command may have executed.
> 			// tryAbandon indicates whether the command is already executing
> ```

why even bother trying to abandon? in the event of failed abandonment, why do we continue to wait for the proposal result?

---

*[pkg/storage/replica.go, line 3143 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoQ44DoIv7_4c4pK2T:-KWoPiB3tJneZBoP3X6A:b-2n0m5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L3143)):*
> ```Go
> 				doneCh := cmd.LocalProposalData.doneCh
> 				cmd.LocalProposalData = innerPD.LocalProposalData
> 				cmd.endCmds = endCmds
> ```

this would be substantially less confusing if it read:
`cmd.LocalProposalData.endCmds = endCmds`
same for the next line.

---

*[pkg/storage/replica.go, line 3145 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoPo_5cVVjz7uSpNcb:-KWoPo_5cVVjz7uSpNcc:b-yn1wlw) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L3145)):*
> ```Go
> 				cmd.endCmds = endCmds
> 				cmd.doneCh = doneCh
> 				cmd.ctx = nil // already have ctx
> ```

not your code, but what is this?

---

*[pkg/storage/replica_test.go, line 5754 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoRDlBcPshoS7tn3vk:-KWoRDlBcPshoS7tn3vl:b2cdkch) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica_test.go#L5754)):*
> ```Go
> // TestReplicaCancelRaft checks that it is possible to safely abandon Raft
> // commands via a cancelable context.Context.
> ```

what's this empty line?

---

*[pkg/storage/replica_test.go, line 5766 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoQy1RbQj2ZE7XuCZF:-KWoQy1RbQj2ZE7XuCZG:b-3n710k) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica_test.go#L5766)):*
> ```Go
> 				cfg.TestingKnobs.TestingCommandFilter =
> 					func(filterArgs storagebase.FilterArgs) *roachpb.Error {
> 						if !filterArgs.Req.Header().Key.Equal(key) {
> ```

would be clearer as
```
if filterArgs.Req.Header().Key.Equal(key) {
  cancel()
}
return nil
```

---

*[pkg/storage/replica_test.go, line 5780 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoRGyThRxJu9ikZ6py:-KWoRGyThRxJu9ikZ6pz:b354p86) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica_test.go#L5780)):*
> ```Go
> 				cancel()
> 				tc.repl.mu.Lock()
> 				tc.repl.mu.submitProposalFn = func(*ProposalData) error {
> ```

what's this?

---

*[pkg/storage/replica_test.go, line 5793 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoRKAh86e63Mu870hf:-KWoRKAh86e63Mu870hg:b-rw5yeh) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica_test.go#L5793)):*
> ```Go
> 			}
> 			br, pErr := tc.repl.addWriteCmd(ctx, ba)
> 			if pErr == nil {
> ```

can this check for a more specific error?

maybe I'm tired, but this is very hard to parse. can it be written as
```
if cancelEary {
  // check the expected outcome
} else {
 // check the expected outcome
}
```
? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 5 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1234 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWqLpxR1UwWfWG_rz5j:-KWqLpxR1UwWfWG_rz5k:b-wdh2sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L1234)):*
> ```Go
> // done removes pending commands from the command queue and updates
> // the timestamp cache using the final timestamp of each command.  The
> // returned error replaces the supplied error.
> ```

What returned error? Looks like this comment rotted in review. 

---

*[pkg/storage/replica.go, line 1800 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWqNpppwMm5Y2cuNJsC:-KWqNpppwMm5Y2cuNJsD:bwbkd4z) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L1800)):*
> ```Go
> 		select {
> 		case propResult := <-ch:
> 			endCmds = nil // these will have been invoked post-Raft.
> ```

Isn't this racy? It's possible that our context is canceled just as the command finishes, so we hit the `<-ctxDone` path instead of this one and both this thread and the raft thread invoke endCmds.done. I think we need a synchronized flag in `endCmds` to ensure `done` is only run once. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/storage/replica.go, line 1234 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWqLpxR1UwWfWG_rz5j:-KWtaYFuXPbhhzjEmn6x:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L1234)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> What returned error? Looks like this comment rotted in review. 
</details>
Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/storage/replica.go, line 1800 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoNCmjK6qOiZPOVOal:-KWtb3xBshJ3QGTxMHLN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L1800)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this comment is anemic, can you flesh it out? In particular, isn't it always the intention that `endCmds` is processed post-Raft?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/storage/replica.go, line 1800 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWqNpppwMm5Y2cuNJsC:-KWtbWPKlSYoptzuextJ:b-coybdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L1800)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Isn't this racy? It's possible that our context is canceled just as the command finishes, so we hit the `<-ctxDone` path instead of this one and both this thread and the raft thread invoke endCmds.done. I think we need a synchronized flag in `endCmds` to ensure `done` is only run once. 
</details>
If the `<-ctxDone` path is hit, either it succeeds in the `tryAbandon`, in which case it relies on the raft thread to invoke `endCmds.done`, or else it fails in the `tryAbandon`, in which case this loop iterates until the command is actually completed. In both cases, only the raft thread will invoke `endCmds.done`.

Am I missing something?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/storage/replica.go, line 1805 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoNmzEruMqPl8UMkmY:-KWteRjVcx4-PFobDgT6:bm8xepc) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L1805)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this comment is confusing to put it mildly.
> 
> The ambiguous result error is only returned if the context was cancelled **and** the command might already be executing, right? The first sentence states that the ambiguous result is returned unconditionally, and then the second sentence adds a condition.
> 
> Also, if the context is cancelled and then we fail to abandon, this loops back around and waits for the result of the raft command, even though the client no longer cares - is that the intention?
> 
> It seems that `endCmds` is being set to nil below to avoid the defer, but again, it deserves explicit commentary. In fact, I would rather we removed that defer completely; the few lines saved are not worth the subtlety and confusion that result from its presence.
</details>
I clarified comment.

Yes, that's the intention. If we can't abandon, then somewhere in the bowels of Raft our context may be in use, so we have to wait it out to avoid panics.

Added commentary on setting endCmds to nil.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/storage/replica.go, line 1817 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoPEWJnbBVgNDk5FAO:-KWtf1Y3Xb9QeD3Du7fJ:bg4nitc) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L1817)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why even bother trying to abandon? in the event of failed abandonment, why do we continue to wait for the proposal result?
</details>
Because of `processRaftCommand`'s use of our context. This whole point of `tryAbandon` is to deal with that case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/storage/replica.go, line 3143 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoQ44DoIv7_4c4pK2T:-KWtc5tX4K4Ad7KtSfLy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L3143)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this would be substantially less confusing if it read:
> `cmd.LocalProposalData.endCmds = endCmds`
> same for the next line.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/storage/replica.go, line 3145 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoPo_5cVVjz7uSpNcb:-KWtc7-PpXy-k5zDkejM:bk1vwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica.go#L3145)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not your code, but what is this?
</details>
Not 100% sure.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/storage/replica_test.go, line 5754 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoRDlBcPshoS7tn3vk:-KWtcH823xIroji75ukd:ba3dfob) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica_test.go#L5754)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's this empty line?
</details>
Removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/storage/replica_test.go, line 5766 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoQy1RbQj2ZE7XuCZF:-KWtcLuXdnA5XeSz66Wb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica_test.go#L5766)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> would be clearer as
> ```
> if filterArgs.Req.Header().Key.Equal(key) {
>   cancel()
> }
> return nil
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/storage/replica_test.go, line 5780 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoRGyThRxJu9ikZ6py:-KWtcaZgDmpgJRbIxQg3:b-w3kj91) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica_test.go#L5780)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's this?
</details>
Meant to skip actually sending something down to Raft.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
*[pkg/storage/replica_test.go, line 5793 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoRKAh86e63Mu870hf:-KWtdExpw17l86hz2aP7:b2f0gba) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica_test.go#L5793)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can this check for a more specific error?
> 
> maybe I'm tired, but this is very hard to parse. can it be written as
> ```
> if cancelEary {
>   // check the expected outcome
> } else {
>  // check the expected outcome
> }
> ```
> ? 
</details>
Not sure what you're asking for here. There's no error to check for...?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Sending various fixes in a new PR.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-:-KWtfKXDQy1_kIwAyXw-:b-smbvtd)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_test.go, line 5793 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772#-KWoRKAh86e63Mu870hf:-KWtfXVE-QumODUPwMoc:b-b6mwto) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a787be607da7bc1d4c8f22ced05af31486fe47/pkg/storage/replica_test.go#L5793)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Not sure what you're asking for here. There's no error to check for...?
</details>
You're failing the test if the error is nil, meaning that you expect an error. I'm suggesting you check explicitly for the error you expect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10772)*
<!-- Sent from Reviewable.io -->
